### PR TITLE
task_k (MVVM)

### PR DESCRIPTION
**Pull request by MaximBadanov**
Пока пытаюсь понять, как все запросы вызывать одним действием.
Убрал весь код из ResultViewViewModel, тк  мы уже не будем делать сетевые запросы там, а будем передавать данные и декодить постеры.

### CODE REVIEW

File: swiftui_project/swiftui_project/GenreScreen/GenreViewViewModel/GenreViewViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import SwiftUI
+import Foundation
+
+class GenreViewViewModel: ObservableObject {
+    @Published var movie: MovieResponseModel?
+    private var subscriber: AnyCancellable?
+    private let dataManager: DataManager
+    
+    init() {
+        guard let network = DIContainer.shared.injectDependency(dependency: NetworkService()) else {
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/GenreScreen/GenreViewViewModel/GenreViewViewModel.swift:11]: You should here take network from DI not inject here

File: swiftui_project/swiftui_project/GenreScreen/GenreViewViewModel/GenreViewViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            fatalError("Service not found in DI container")
+        }
+        dataManager = DataManager(network: network)
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/GenreScreen/GenreViewViewModel/GenreViewViewModel.swift:14]: It should be in DI

File: swiftui_project/swiftui_project/GenreScreen/GenresScreen.swift [hunk 3]
```diff
@@ -13,17 +15,31 @@ struct GenresScreen: View {
                     weight: .bold
                 )
             GenresViewWithScroll()
+                .onAppear {
+                    viewModel.fetchMoviesByGenre(
+                        requestModel: RequestModel(
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/GenreScreen/GenresScreen.swift:20]: You must not create Request on view side

File: swiftui_project/swiftui_project/GenreScreen/GenresScreen.swift [hunk 4]
```diff
@@ -1,10 +1,12 @@
 import SwiftUI
 
-
 struct GenresScreen: View {
+    @ObservedObject private var viewModel = GenreViewViewModel()
+    @State private var genres: [String] = ["27", "12"]
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/GenreScreen/GenresScreen.swift:5]: The genres should be chosen by users not hardcoded

File: swiftui_project/swiftui_project/GenreScreen/GenresScreen.swift [hunk 5]
```diff
@@ -13,17 +15,31 @@ struct GenresScreen: View {
...
+                        genres: genres
+                    )
+                }
             Spacer(minLength: UISize.size8)
             HStack(spacing: UISize.size8) {
                 Spacer(minLength: UISize.size8)
                     .padding(.horizontal, UISize.size24)
                 NavigationLink("Continue",
-                               destination: ResultView())
+                               destination: ResultView(
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/GenreScreen/GenresScreen.swift:34]: I think better fetch random movie on result view where should be movie

File: swiftui_project/swiftui_project/NetworkService/NetworkServiceSupport/Urls.swift [hunk 6]
```diff
@@ -1,3 +1,5 @@
 enum Urls: String {
     case moviedbGenres = "https://api.themoviedb.org/3/genre/movie/list?language=en"
+    case moviesByGenres = "https://api.themoviedb.org/3/discover/movie?&with_genres="
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/NetworkService/NetworkServiceSupport/Urls.swift:3]: &with_genres should be hardcoded, you should send url/body parameters in network service from request model

File: swiftui_project/swiftui_project/ResultView/MovieView.swift [hunk 7]
```diff
@@ -1,11 +1,11 @@
...
 struct MovieView: View {
-    let jokerMovie = Movie(
-        movieTitle: "Joker" ,
-        genre: "Thriller",
-        posterName: "joker"
-    )
+    @State var title: String
+    @State var poster: String
+    
+    private let baseURL = "https://image.tmdb.org/t/p/"
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/ResultView/MovieView.swift:7]: Nothing for network or business logic shouldn’t be on view side

**Review by iosnickse — CHANGES_REQUESTED** [review]

File: swiftui_project/swiftui_project/GenreScreen/GenreViewViewModel/GenreViewViewModel.swift [hunk 8]
```diff
@@ -30,6 +32,7 @@ extension GenreViewViewModel: GenreViewViewModelProtocol {
                 guard let movie = data.results.randomElement() else { return }
                 print("Movie: \(movie.title ), Poster: \(movie.poster )")
                 self.movie = MovieResponseModel(title: movie.title , poster: movie.poster )
+                self.resultViewModel.movie = movie
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/GenreScreen/GenreViewViewModel/GenreViewViewModel.swift:35]: You shouldn't pass one vm to another VM

Comment (iosnickse) [reply swiftui_project/swiftui_project/GenreScreen/GenreViewViewModel/GenreViewViewModel.swift:35]: If you want to send data between screen, for simple way (just to show) you must use just let property in view, new input will redraw view and that's it, even without VM

**Review by iosnickse — CHANGES_REQUESTED** [review]

File: swiftui_project/swiftui_project/GenreScreen/GenreButton.swift [hunk 9]
```diff
@@ -3,15 +3,33 @@ import SwiftUI
...
+        self._id = State(initialValue: id)
+    }
+    
     var body: some View {
         VStack(spacing: UISize.size8) {
             Button(emoji) {
                 isSelected.toggle()
+                
+                if isSelected {
+                    if !GenreButtonViewModel.shared.genres.contains(id) {
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/GenreScreen/GenreButton.swift:25]: All logic must be in vm

File: swiftui_project/swiftui_project/ResultView/MovieView.swift [hunk 10]
```diff
@@ -39,7 +39,7 @@ struct MovieView: View {
                     weight: .heavy
                 )
                 .padding(.horizontal, UISize.size16)
-            Text("Genre")
+            Text(GenreButtonViewModel.shared.genres[0])
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/ResultView/MovieView.swift:42]: Don’t use indexes by hardcode

File: swiftui_project/swiftui_project/Supports&Extensions/Genre/AllGenres.swift [hunk 11]
```diff
@@ -28,115 +28,134 @@ enum AllGenres: CaseIterable {
...
+                color: .yellow,
+                id: "10402"
             )
         case .mystery:
             return Genre(
                 title: "Mystery",
                 emoji: "👻",
-                color: .gray
+                color: .gray,
+                id: "9648"
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/Supports&Extensions/Genre/AllGenres.swift:123]: Don’t use hardcode here

File: swiftui_project/swiftui_project/Supports&Extensions/Genre/AllGenres.swift [hunk 12]
```diff
@@ -28,115 +28,134 @@ enum AllGenres: CaseIterable {
...
             )
         case .romance:
             return Genre(
                 title: "Romance",
                 emoji: "🌹",
-                color: .indigo
+                color: .indigo,
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/Supports&Extensions/Genre/AllGenres.swift:129]: Better to fetch genres from api and create model for it or optional init by title or something like that, my opinion that you are good to refactor the enum to model codable and fetch from API

File: swiftui_project/swiftui_project/GenreScreen/GenreButton.swift [hunk 13]
```diff
@@ -3,15 +3,33 @@ import SwiftUI
...
+                        GenreButtonViewModel.shared.genres.append(id)
+                        print(GenreButtonViewModel.shared.genres)
```
[the remark this answers is not in the document]
Comment (MaximBadanov) [reply swiftui_project/swiftui_project/GenreScreen/GenreButton.swift:27]: это для проверки работы логики, пусть пока будет

**Pull request closed without merging**